webkit2/webkit2.uri-scheme-request.lisp

```diff
@@ -0,0 +1,103 @@
+
+(defvar +webkit-plugin-error-connection-cancelled+ 4
+  "A separate value for the connection cancelled plugin error for use
+  in WebKitURISchemeRequest callbacks.")
+
+(defcfun "webkit_uri_scheme_request_get_scheme" :string
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-scheme)
+
+(defcfun "webkit_uri_scheme_request_get_uri" :string
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-uri)
+
+(defcfun "webkit_uri_scheme_request_get_path" :string
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-path)
+
+(defcfun "webkit_uri_scheme_request_get_web_view" (g-object webkit-web-view)
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-web-view)
+
+(defcfun "webkit_uri_scheme_request_finish" :void
+  (request (g-object webkit-uri-scheme-request))
+  (stream (g-object g-memory-input-stream))
+  (stream-length :long)
+  (content-type :string))
+(export 'webkit-uri-scheme-request-finish)
+
+(defcfun ("webkit_uri_scheme_request_finish_error" %webkit-uri-scheme-request-finish-error) :void
+  (request (g-object webkit-uri-scheme-request))
+  (g-error :pointer))
+
+(defun webkit-uri-scheme-request-finish-error (request error-string)
+  (%webkit-uri-scheme-request-finish-error
+   request (glib::%g-error-new-literal
+            +webkit-plugin-error+
+            +webkit-plugin-error-connection-cancelled+
+            error-string)))
+(export 'webkit-uri-scheme-request-finish-error)
+
+(cffi:defcallback uri-scheme-processed :void ((request (g-object webkit-uri-scheme-request))
+                                              (user-data :pointer))
+  (let ((callback (find (cffi:pointer-address user-data) callbacks :key (function callback-id))))
+    (when (callback-function callback)
+      ;; Callback function returns data-string as a first value
+      ;; and data type (e.g., "text/html") as an optional second
+      ;; value.
+      (destructuring-bind (data &optional (data-type "text/html"))
+          (multiple-value-list (funcall (callback-function callback) request))
+        (handler-case
+            (multiple-value-bind (ffi-string ffi-string-length)
+                (cffi:foreign-string-alloc data)
+              (unwind-protect
+                   (let* ((stream (g-memory-input-stream-new-from-data
+                                   ffi-string -1 ; -1 is for auto-detection based on NULL character
+                                   (callback g-notify-destroy-null))))
+                     (webkit-uri-scheme-request-finish request stream ffi-string-length data-type)
+                     (gobject:g-object-unref (pointer stream)))
+                (cffi:foreign-string-free ffi-string)))
+          (error (c)
+            (webkit-uri-scheme-request-finish-error
+             request (format nil "The custom url request for URI ~a failed with ~a: ~a"
```

Comment:
I don't understand this sentence. Shouldn't it be

"The custom request for URI ~a failed"

?

Reply:
It could but I needet it to be more verbose about errors, listing both it's type and error message. Still looks alright to me — it gives more details.

Reply:
Misunderstanding: I didn't mean to remove that last `~a`, only to remove "url" after "custom".